### PR TITLE
[FIX] account: restrict to create payment without payment date

### PR DESCRIPTION
**Pull request by alsh-odoo**
This traceback raises when user tries to create payment without payment date.

To reproduce this issue:

1) Install `account`
2) Open any existing invoice (not a DRAFT one)
3) Click `REGISTER PAYMENT` button.
4) Now remove `Payment Date` in payment wizard.

Error: 
```
ValueError: not enough values to unpack (expected 1, got 0)
  File "odoo/models.py", line 5417, in ensure_one
    _id, = self._ids
ValueError: Expected singleton: account.payment.term()
  File "odoo/http.py", line 2134, in __call__
    response = request._serve_db()
  File "odoo/http.py", line 1710, in _serve_db
    return service_model.retrying(self._serve_ir_http, self.env)
  File "odoo/service/model.py", line 133, in retrying
    result = func()
  File "odoo/http.py", line 1737, in _serve_ir_http
    response = self.dispatcher.dispatch(rule.endpoint, args)
  File "odoo/http.py", line 1938, in dispatch
    result = self.request.registry['ir.http']._dispatch(endpoint)
  File "odoo/addons/base/models/ir_http.py", line 191, in _dispatch
    result = endpoint(**request.params)
  File "odoo/http.py", line 717, in route_wrapper
    result = endpoint(self, *args, **params_ok)
  File "addons/web/controllers/dataset.py", line 30, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "addons/web/controllers/dataset.py", line 26, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "odoo/api.py", line 461, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "odoo/api.py", line 448, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "odoo/models.py", line 6744, in onchange
    todo = [
  File "odoo/models.py", line 6747, in <listcomp>
    if name not in done and snapshot0.has_changed(name)
  File "odoo/models.py", line 6544, in has_changed
    return self[name] != record[name]
  File "odoo/models.py", line 6171, in __getitem__
    return self._fields[key].__get__(self, type(self))
  File "odoo/fields.py", line 1156, in __get__
    self.recompute(record)
  File "odoo/fields.py", line 1366, in recompute
    apply_except_missing(self.compute_value, recs)
  File "odoo/fields.py", line 1339, in apply_except_missing
    func(records)
  File "odoo/fields.py", line 1388, in compute_value
    records._compute_field_value(self)
  File "odoo/models.py", line 4524, in _compute_field_value
    fields.determine(field.compute, self)
  File "odoo/fields.py", line 100, in determine
    return needle(*args)
  File "addons/account/wizard/account_payment_register.py", line 558, in _compute_amount
    wizard.amount = wizard._get_total_amount_in_wizard_currency_to_full_reconcile(batch_result)[0]
  File "addons/account/wizard/account_payment_register.py", line 524, in _get_total_amount_in_wizard_currency_to_full_reconcile
    return self._get_total_amount_using_same_currency(batch_result, early_payment_discount=early_payment_discount)
  File "addons/account/wizard/account_payment_register.py", line 506, in _get_total_amount_using_same_currency
    amount += move.invoice_payment_term_id._get_amount_due_after_discount(move.amount_total, move.amount_tax)#todo currencies
  File "addons/account/models/account_payment_term.py", line 55, in _get_amount_due_after_discount
    self.ensure_one()
  File "odoo/models.py", line 5420, in ensure_one
    raise ValueError("Expected singleton: %s" % self)
```

Note:- 'TypeError: unsupported operand type(s) for +=: 'float' and 'NoneType'' is also produced, if the invoice is having
`invoice_payment_term_id`

When user removes `payment_date` a  `_compute_amount` method will be called, In which
`_get_total_amount_in_wizard_currency_to_full_reconcile` method is used.

In `_get_total_amount_in_wizard_currency_to_full_reconcile` method,
`_get_total_amount_using_same_currency` is used to get total amount.

On `_get_total_amount_using_same_currency` method `payment_date` is passed as an argument for `_is_eligible_for_early_payment_discount` method.

See: https://github.com/odoo/odoo/blob/ed2e26633458f1284437286f71eba43c8d1818a7/addons/account/wizard/account_payment_register.py#L490-L498

On `_is_eligible_for_early_payment_discount` method if no `reference date`, its returning `True`. 
which leads to above traceback.

See:
https://github.com/odoo/odoo/blob/ed2e26633458f1284437286f71eba43c8d1818a7/addons/account/models/account_move.py#L1902-L1905

sentry-4337953422

### CODE REVIEW

**Comment by alsh-odoo** [comment]
Hello @odoo/sentry_reviewers 

Please check this PR

Thank You !!

File: addons/account/models/account_move.py [hunk 1]
```diff
@@ -1902,7 +1902,7 @@ def _validate_taxes_country(self):
     def _is_eligible_for_early_payment_discount(self, currency, reference_date):
         self.ensure_one()
         if not reference_date:
-            return True
+            return None
```
Comment (JulienAlardot) [inline addons/account/models/account_move.py:1905]:
```suggestion
            return
```
return is None by default

**Review by JulienAlardot — CHANGES_REQUESTED** [review]
Tested functionally and it does the job, small remark though

File: addons/account/models/account_move.py (hunk 1 above)
Comment (alsh-odoo) [reply addons/account/models/account_move.py]: Hello @JulienAlardot 
Changes are done.
Please check Thank You!!

**Comment by JulienAlardot** [comment]
LGTM @jbw-odoo

**Comment by fpodoo** [comment]
That does not seem good to me; if no payment date is provided in a payment, it means "asap", that can be estimated to today. You can not consider that `_is_eligible_for_early_payment_discount` return True if no payment date is provided; if no payment date, you should compute with today as the date.

**Comment by alsh-odoo** [comment]
Hello @smetl 
Changes are done. Please review 
Thanks!!

File: addons/account/wizard/account_payment_register.py [hunk 2]
```diff
@@ -554,7 +556,9 @@ def _compute_amount(self):
     @api.depends('can_edit_wizard', 'payment_date', 'currency_id', 'amount')
     def _compute_early_payment_discount_mode(self):
         for wizard in self:
-            if wizard.can_edit_wizard and wizard.currency_id:
+            if not wizard.journal_id or not wizard.currency_id or not wizard.payment_date:
+                wizard.early_payment_discount_mode = wizard.early_payment_discount_mode
+            elif wizard.can_edit_wizard and wizard.currency_id:
```
Comment (smetl) [inline addons/account/wizard/account_payment_register.py:561]:
```suggestion
            elif wizard.can_edit_wizard:
```

**Pull request closed without merging**